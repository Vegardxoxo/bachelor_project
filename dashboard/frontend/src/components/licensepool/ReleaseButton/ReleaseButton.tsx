import React from 'react';
import {Button} from '@mui/material';
import {useRecoilValue, useSetRecoilState} from "recoil";
import {refreshTableAtom, userAtom} from "../../../globalVariables/variables";

type ReserveButtonProps = {
    spc_id: number;
    primary_user_email: string;
    application_name: string;
    organization: string;
    price: number;

}

/**
 * A button that can be used to release licenses.
 * @param spc_id The id of the license to release
 * @param primary_user_email The email of the user who is requesting to free the license.
 * @param application_name The name of the application
 * @param organization The name of the organization
 * @param price The price of the license
 */
const ReleaseButton: React.FC<ReserveButtonProps> = ({
                                                         spc_id,
                                                         primary_user_email,
                                                         application_name,
                                                         organization,
                                                         price
                                                     }) => {
    const accessToken = localStorage.getItem('access');
    const userInfo = useRecoilValue((userAtom))
    const isUnitHead = userInfo.is_unit_head;
    const setRefresh = useSetRecoilState(refreshTableAtom)

    // Handle button click based on whether the user is a unit head
    const handleClick = async () => {
        if (!isUnitHead) {
            const data = await requestFreeOwnLicense();
            if (data) {
                alert('Forespørsel sendt sendt til lisensansvarlig!')
            }
        } else {
            const data = await freeLicense();
            if (data) {
                alert('Lisens frigjort!')
            }

        }
    }

    // Send a request to free the user's own license
    const requestFreeOwnLicense = async () => {
        try {
            const response = await fetch('http://127.0.0.1:8000/api/requests/create', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${accessToken}`,
                },
                body: JSON.stringify({
                    'contact_organization': organization,
                    'application_name': application_name,
                    'request': 'add',
                    'requested_by': primary_user_email,
                    'price': price,
                    'spc_id': spc_id,
                }),
            });
            const data = await response.json();
            if (response.ok) {
                return data;
            } else {
                alert(data.non_field_errors[0])
            }
        } catch (error) {
            console.error(error);
        }
    };


    // Free the license and update the pool
    const freeLicense = async () => {
        try {
            const response = await fetch('http://127.0.0.1:8000/api/pool/create', {
                method: 'POST', // Change this from 'GET' to 'POST'
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${accessToken}`,
                },
                body: JSON.stringify({
                    'freed_by_organization': organization,
                    'application_name': application_name,
                    'price': price,
                    'spc_id': spc_id,
                }),
            });
            const data = await response.json();
            if (response.ok) {
                setRefresh((old) => !old)
                return data;
            } else {
                alert(data.non_field_errors[0])
            }
        } catch (error) {
            console.error(error);
        }
    };


    // Render the button with an onClick handler
    return (
        <Button data-testid="foresporsel-release-id" onClick={handleClick} variant="contained"
                sx={{backgroundColor: '#00953b'}}>
            {isUnitHead ? 'Legg i lisensportalen' : 'Send forespørsel'}
        </Button>
    );

}
export default ReleaseButton;