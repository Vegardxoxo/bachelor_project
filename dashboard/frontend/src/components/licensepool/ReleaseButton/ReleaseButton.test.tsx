import React from 'react';
import '@testing-library/jest-dom';
import {cleanup, fireEvent, render, screen, waitFor} from '@testing-library/react';
import ReleaseButton from './ReleaseButton';
import * as recoil from "recoil";
import {RecoilRoot} from "recoil";
import 'isomorphic-fetch';
import renderer from "react-test-renderer";

const mockProps = {
    "id": 1,
    "full_name": "Jannik Georg Solvang",
    "primary_user_email": "kekw",
    'organization': 'fisk',
    'application_name': 'APSIS Pro [Web]',
    'price': 1000

}


describe('The button for regular users', () => {
    beforeEach(() => {
        jest.spyOn(global, 'fetch').mockImplementation(() =>
            Promise.resolve(new Response(JSON.stringify({})))
        );
        render(
            <RecoilRoot> <ReleaseButton spc_id={mockProps.id} primary_user_email={mockProps.full_name}
                                        application_name={mockProps.application_name}
                                        organization={mockProps.organization}
                                        price={mockProps.price}/></RecoilRoot>);
    });
    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
    })


    it('renders without crashing', async () => {
        expect(screen.getByText('Send forespørsel')).toBeInTheDocument();
    })

    it('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <RecoilRoot> <ReleaseButton spc_id={mockProps.id} primary_user_email={mockProps.full_name}
                                        application_name={mockProps.application_name}
                                        organization={mockProps.organization}
                                        price={mockProps.price}/></RecoilRoot>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    })

    it('can be clicked', async () => {
        const alertSpy = jest.spyOn(window, 'alert').mockImplementation(() => {//check
        });
        fireEvent.click(screen.getByText('Send forespørsel'));
        await waitFor(() => expect(alertSpy).toHaveBeenCalledWith('Forespørsel sendt sendt til lisensansvarlig!'));
        alertSpy.mockRestore();
    })

})

describe('The button for unit heads', () => {
    beforeEach(() => {
        jest.spyOn(global, 'fetch').mockImplementation(() =>
            Promise.resolve(new Response(JSON.stringify({})))
        );
        jest.spyOn(recoil, 'useRecoilValue').mockReturnValue({is_unit_head: true});
        render(
            <RecoilRoot> <ReleaseButton spc_id={mockProps.id} primary_user_email={mockProps.full_name}
                                        application_name={mockProps.application_name}
                                        organization={mockProps.organization}
                                        price={mockProps.price}/></RecoilRoot>);

    })

    afterEach(() => {
        cleanup();
        jest.restoreAllMocks();
    })

    it('renders with "Legg i lisensportalen" text', async () => {
        expect(screen.getByText('Legg i lisensportalen')).toBeInTheDocument();
    })

    it('can be clicked with "Legg i lisensportalen" text', async () => {
        const alertSpy = jest.spyOn(window, 'alert').mockImplementation(() => {
            // do nothing
        });
        fireEvent.click(screen.getByText('Legg i lisensportalen'));
        await waitFor(() => expect(alertSpy).toHaveBeenCalledWith('Lisens frigjort!'));
        alertSpy.mockRestore();
    })
})
