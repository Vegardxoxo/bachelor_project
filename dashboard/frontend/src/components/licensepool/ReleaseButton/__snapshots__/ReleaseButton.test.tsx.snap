// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The button for regular users matches snapshot 1`] = `
Array [
  " ",
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-2xjwzt-MuiButtonBase-root-MuiButton-root"
    data-testid="foresporsel-release-id"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Send foresp√∏rsel
  </button>,
]
`;
